### PR TITLE
Exploit P10 expand mask instructions in existing and new operations.

### PR DESCRIPTION
**Pull request by munroesj52**
Use P10's expand mask instructions for set bool (setb) from sign for
P10 specific implementations of existing and new vec_setb_* operations.
This includes the existing vec_setb_sq (__int128) and vec_setb_qp
(float128) operations, plus new operations for integer signed
long long, int, short and char and floating-point single and double.

	* src/pveclib/vec_char_ppc.h (vec_setb_sb): New operation.
	* src/pveclib/vec_f128_ppc.h (vec_setb_qp [_ARCH_PWR10]):
	Use vexpandqm for POWER10.
	* src/pveclib/vec_f32_ppc.h (vec_setb_sp): New operation.
	* src/pveclib/vec_f64_ppc.h (vec_setb_dp): New operation.
	* src/pveclib/vec_int128_ppc.h (vec_setb_sq [_ARCH_PWR10]):
	Use vexpandqm for POWER10.
	* src/pveclib/vec_int16_ppc.h (vec_setb_sh): New operation.
	* src/pveclib/vec_int32_ppc.h (vec_setb_sw): New operation.
	* src/pveclib/vec_int64_ppc.h (vec_setb_sd): New operation.

	* src/testsuite/arith128_test_char.c (test_setbb):
	New unit test.
	(test_vec_char): Add test_setbb to driver function.

	* src/testsuite/arith128_test_f32.c (test_setb_sp):
	New unit test.
	(test_vec_f32): Add test_setb_sp to driver function.

	* src/testsuite/arith128_test_f64.c (test_setb_dp):
	New unit test.
	(test_vec_f64): Add test_setb_dp to driver function.

	* src/testsuite/arith128_test_i16.c  (test_setbh):
	New unit test.
	(test_vec_i16): Add test_setbh to driver function.

	* src/testsuite/arith128_test_i32.c  (test_setbw):
	New unit test.
	(test_vec_i32): Add test_setbw to driver function.

	* src/testsuite/arith128_test_i64.c  (test_setbd):
	New unit test.
	(test_vec_i64): Add test_setdh to driver function.

	* src/testsuite/vec_char_dummy.c (test_setb_sb)
	New compile test.
	* src/testsuite/vec_f128_dummy.c [PVECLIB_DISABLE_F128MATH]
	Add test_scalar_add128 and test_scalar_mul128.
	(test_setb_qp): New compile test.
	* src/testsuite/vec_f32_dummy.c (__test_setb_sp)
	New compile test.
	* src/testsuite/vec_f64_dummy.c (__test_setb_dp)
	New compile test.
	* src/testsuite/vec_int16_dummy.c (test_setb_sh)
	New compile test.
	* src/testsuite/vec_int32_dummy.c (test_setb_sw)
	New compile test.
	* src/testsuite/vec_int64_dummy.c (test_setb_sd)
	New compile test.

	src/testsuite/vec_pwr10_dummy.c (test_setb_qp_PWR10,
	test_setb_sb_PWR10, test_setb_sh_PWR10, test_setb_sw_PWR10,
	test_setb_sd_PWR10): New compile test.
	src/testsuite/vec_pwr9_dummy.c (test_setb_qp_PWR9,
	test_setb_sb_PWR9, test_setb_sh_PWR9, test_setb_sw_PWR9,
	test_setb_sd_PWR9): New compile test.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

### CODE REVIEW

File: src/pveclib/vec_char_ppc.h [hunk 1]
```diff
@@ -753,6 +753,41 @@ vec_popcntb (vui8_t vra)
...
+ */
+
+static inline vb8_t
+vec_setb_sb (vi8_t vra)
+{
+  vb8_t result;
+
+#if defined (_ARCH_PWR10)  && (__GNUC__ >= 10)
+  __asm__(
+      "vexpandbm %0,%1;\n"
```
Comment (ThinkOpenly) [inline src/pveclib/vec_char_ppc.h:779]: You don't need the trailing `;\n`

File: src/pveclib/vec_f128_ppc.h [hunk 2]
```diff
@@ -1132,20 +1139,31 @@ vec_iszerof128 (__binary128 f128)
 static inline vb128_t
 vec_setb_qp (__binary128 f128)
 {
+  vb128_t result;
+#if defined (_ARCH_PWR10)  && (__GNUC__ >= 10)
+  vui128_t vra = vec_xfer_bin128_2_vui128t (f128);
+  __asm__(
+      "vexpandqm %0,%1;\n"
```
Comment (ThinkOpenly) [inline src/pveclib/vec_f128_ppc.h:1146]: `;\n` here, too

File: src/pveclib/vec_f128_ppc.h [hunk 3]
```diff
@@ -1132,20 +1139,31 @@ vec_iszerof128 (__binary128 f128)
...
+      : "=v" (result)
+      : "v" (vra)
+      : );
+#else
 #if defined (_ARCH_PWR9) && defined (scalar_test_neg) && (__GNUC__ > 7)
```
Comment (ThinkOpenly) [inline src/pveclib/vec_f128_ppc.h:1151]: you could merge these into an `#elif`

File: src/pveclib/vec_f32_ppc.h [hunk 4]
```diff
@@ -1090,6 +1090,34 @@ vec_iszerof32 (vf32_t vf32)
...
+/*! \brief Vector Set Bool from Sign, Single Precision.
+ *
+ *  For each float, propagate the sign bit to all 32-bits of that
+ *  word. The result is vector bool int reflecting the sign
+ *  bit of each 32-bit float.
+ *
+ *  The resulting mask can be used in masking and select operations.
+ *
+ *  \note This operation will set the sign mask regardless of data
+ *  class. while the Vector Test Data Class will not distinguish
```
Comment (ThinkOpenly) [inline src/pveclib/vec_f32_ppc.h:1102]: s/`.`/`,`/ ?

File: src/pveclib/vec_int128_ppc.h [hunk 5]
```diff
@@ -5755,10 +5755,21 @@ vec_setb_ncq (vui128_t vcy)
 static inline vb128_t
 vec_setb_sq (vi128_t vra)
 {
+  vui128_t result;
+
+#if defined (_ARCH_PWR10)  && (__GNUC__ >= 10)
+  __asm__(
+      "vexpandqm %0,%1;\n"
```
Comment (ThinkOpenly) [inline src/pveclib/vec_int128_ppc.h:5762]: `;\n` here, too

File: src/pveclib/vec_char_ppc.h [hunk 6]
```diff
@@ -753,6 +753,41 @@ vec_popcntb (vui8_t vra)
 #define vec_popcntb __builtin_vec_vpopcntb
 #endif
 
+/*! \brief Vector Set Bool from Signed Byte.
+ *
+ *  For each byte, propagate the sign bit, to all 8-bits of that
```
Comment (ThinkOpenly) [inline src/pveclib/vec_char_ppc.h:758]: s/propagate the sign bit, to/propagate the sign bit to/
this is a global comment, as there are more instances of this below...

File: src/pveclib/vec_int64_ppc.h [hunk 7]
```diff
@@ -2806,6 +2806,50 @@ static inline vui64_t vec_vsrd (vui64_t vra, vui64_t vrb);
...
+  vb64_t result;
+
+#if defined (_ARCH_PWR10)  && (__GNUC__ >= 10)
+  __asm__(
+      "vexpanddm %0,%1;\n"
+      : "=v" (result)
+      : "v" (vra)
+      : );
+#else
+#ifdef _ARCH_PWR8
```
Comment (ThinkOpenly) [inline src/pveclib/vec_int64_ppc.h:2837]: could be combined into `#elif`

**Review by ThinkOpenly — APPROVED** [review]
tiny, ignorable nits.

File: src/pveclib/vec_char_ppc.h (hunk 1 above)
Comment (munroesj52) [reply src/pveclib/vec_char_ppc.h:779]: OK

File: src/pveclib/vec_f128_ppc.h (hunk 2 above)
Comment (munroesj52) [reply src/pveclib/vec_f128_ppc.h:1146]: OK

File: src/pveclib/vec_f128_ppc.h (hunk 3 above)
Comment (munroesj52) [reply src/pveclib/vec_f128_ppc.h:1151]: OK

File: src/pveclib/vec_f32_ppc.h (hunk 4 above)
Comment (munroesj52) [reply src/pveclib/vec_f32_ppc.h:1102]: OK

File: src/pveclib/vec_char_ppc.h (hunk 6 above)
Comment (munroesj52) [reply src/pveclib/vec_char_ppc.h:758]: OK

File: src/pveclib/vec_int64_ppc.h (hunk 7 above)
Comment (munroesj52) [reply src/pveclib/vec_int64_ppc.h:2837]: OK

**Pull request merged**